### PR TITLE
feat(smn): support policies in topic

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/smn' TESTARGS='-run TestAccSMNV2Topic_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/smn -v -run TestAccSMNV2Topic_basic -timeout 360m -parallel 4
=== RUN   TestAccSMNV2Topic_basic
=== PAUSE TestAccSMNV2Topic_basic
=== CONT  TestAccSMNV2Topic_basic
--- PASS: TestAccSMNV2Topic_basic (27.26s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/smn       27.362s

make testacc TEST='./huaweicloud/services/acceptance/smn' TESTARGS='-run TestAccSMNV2Topic_policies'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/smn -v -run TestAccSMNV2Topic_policies -timeout 360m -parallel 4
=== RUN   TestAccSMNV2Topic_policies
=== PAUSE TestAccSMNV2Topic_policies
=== CONT  TestAccSMNV2Topic_policies
--- PASS: TestAccSMNV2Topic_policies (35.37s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/smn       35.466s
```

### CODE REVIEW

File: huaweicloud/services/acceptance/smn/resource_huaweicloud_smn_topic_test.go [hunk 1]
```diff
@@ -136,3 +194,36 @@ resource "huaweicloud_smn_topic" "topic_1" {
...
+  introduction             = "created by terraform update"
+}
+`, rName, rName)
+}
+
+func testAccSMNV2TopicConfig_policies_step3(rName string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_smn_topic" "topic_1" {
+  name                     = "%s"
+  display_name             = "The display name of %s"
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/smn/resource_huaweicloud_smn_topic_test.go]: style needs updating

Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/smn/resource_huaweicloud_smn_topic_test.go]: fixed

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/smn' TESTARGS='-run TestAccSMNV2Topic_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/smn -v -run TestAccSMNV2Topic_basic -timeout 360m -parallel 4
=== RUN   TestAccSMNV2Topic_basic
=== PAUSE TestAccSMNV2Topic_basic
=== CONT  TestAccSMNV2Topic_basic
--- PASS: TestAccSMNV2Topic_basic (28.24s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/smn       28.337s

make testacc TEST='./huaweicloud/services/acceptance/smn' TESTARGS='-run TestAccSMNV2Topic_policies'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/smn -v -run TestAccSMNV2Topic_policies -timeout 360m -parallel 4
=== RUN   TestAccSMNV2Topic_policies
=== PAUSE TestAccSMNV2Topic_policies
=== CONT  TestAccSMNV2Topic_policies
--- PASS: TestAccSMNV2Topic_policies (34.73s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/smn       34.818s
```

File: huaweicloud/services/smn/resource_huaweicloud_smn_topic.go [hunk 2]
```diff
@@ -208,6 +277,93 @@ func resourceTopicUpdate(ctx context.Context, d *schema.ResourceData, meta inter
...
+			},
+			"Action": []interface{}{
+				"SMN:Publish",
+				"SMN:QueryTopicDetail",
+			},
+			"Resource": d.Id(),
+		})
+	}
+
+	if len(statement) == 0 {
```
Comment (ShiChangkuo) [inline huaweicloud/services/smn/resource_huaweicloud_smn_topic.go:349]: if both `users_publish_allowed` and `services_publish_allowed` are updated to empty, we will return empty and don‘t call the API

File: docs/resources/smn_topic.md [hunk 3]
```diff
@@ -29,6 +43,16 @@ The following arguments are supported:
 * `display_name` - (Optional, String) Specifies the topic display name, which is presented as the name of the email
   sender in an email message. The name can contains of 0 to 192 characters.
 
+* `users_publish_allowed` - (Optional, String) Specifies the users who can publish messages to this topic.
+  The value can be **\*** which indicates all users or user account URNs separated by comma(,). If left empty,
```
Comment (ShiChangkuo) [inline docs/resources/smn_topic.md]: please give more details about the format:
` The format is urn:csp:iam::domainId:root or *. domainId indicates the account ID of another user, and * indicates the statement is applied for all users.`

https://support.huaweicloud.com/intl/en-us/api-smn/ListTopicAttributes.html#ListTopicAttributes__response_AccessPolicy

File: huaweicloud/services/smn/resource_huaweicloud_smn_topic.go [hunk 4]
```diff
@@ -208,6 +277,93 @@ func resourceTopicUpdate(ctx context.Context, d *schema.ResourceData, meta inter
 	return resourceTopicRead(ctx, d, meta)
 }
 
+func updatePolicies(client *golangsdk.ServiceClient, d *schema.ResourceData, id string) error {
+	if d.HasChanges("users_publish_allowed", "services_publish_allowed") {
+		value, err := buildUpdateAccessPolicy(d)
+		if err != nil {
+			return nil
```
Comment (ShiChangkuo) [inline huaweicloud/services/smn/resource_huaweicloud_smn_topic.go]: should return err or nil?

Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/resource_huaweicloud_smn_topic.go]: should return err, fixed

File: docs/resources/smn_topic.md (hunk 3 above)
Comment (Jason-Zhang9309) [reply docs/resources/smn_topic.md]: sure, the description updated

**Pull request merged**